stl/inc/atomic

```diff
@@ -131,6 +131,74 @@ _NODISCARD extern "C" bool __cdecl __std_atomic_has_cmpxchg16b() noexcept;
 #define ATOMIC_LLONG_LOCK_FREE    2
 #define ATOMIC_POINTER_LOCK_FREE  2
 
+#if _HAS_CXX20
+extern "C" {
+void _CRT_SATELLITE_1 __stdcall __std_atomic_wait_direct(const void* const _Storage, const void* const _Comparand,
+    const std::size_t _Size, unsigned long long& _Wait_context) noexcept;
+void _CRT_SATELLITE_1 __stdcall __std_atomic_notify_one_direct(const void* const _Storage) noexcept;
+void _CRT_SATELLITE_1 __stdcall __std_atomic_notify_all_direct(const void* const _Storage) noexcept;
+
+void _CRT_SATELLITE_1 __stdcall __std_atomic_unwait_direct(
+    const void* const _Storage, unsigned long long& _Wait_context) noexcept;
+
+void _CRT_SATELLITE_1 __stdcall __std_atomic_wait_indirect(
+    const void* const _Storage, unsigned long long& _Wait_context) noexcept;
+void _CRT_SATELLITE_1 __stdcall __std_atomic_notify_one_indirect(const void* const _Storage) noexcept;
+void _CRT_SATELLITE_1 __stdcall __std_atomic_notify_all_indirect(const void* const _Storage) noexcept;
+
+void _CRT_SATELLITE_1 __stdcall __std_atomic_unwait_indirect(
+    const void* const _Storage, unsigned long long& _Wait_context) noexcept;
+
+std::size_t _CRT_SATELLITE_1 __stdcall __std_atomic_get_spin_count(const bool _Is_direct) noexcept;
+
+enum _Atomic_spin_phase : unsigned long long {
+    _Atomic_wait_phase_mask            = 0x0000'000F,
+    _Atomic_spin_value_mask            = 0xFFFF'FFF0,
+    _Atomic_spin_value_step            = _Atomic_wait_phase_mask + 1,
+    _Atomic_counter_value_mask         = ~_Atomic_wait_phase_mask,
+    _Atomic_counter_value_step         = _Atomic_wait_phase_mask + 1,
+    _Atomic_wait_phase_init_spin_count = 0x0000'0000,
+    _Atomic_wait_phase_spin            = 0x0000'0008,
+    _Atomic_wait_phase_wait_locked     = 0x0000'0001,
+    _Atomic_wait_phase_wait_none       = 0x0000'0002,
+    _Atomic_wait_phase_wait_counter    = 0x0000'0004,
+    _Atomic_unwait_needed              = _Atomic_wait_phase_wait_locked,
+};
+
+#if defined(_WIN32_WINNT) && (_WIN32_WINNT >= 0x0602) || defined(_M_ARM) || defined(_M_ARM64)
+inline constexpr bool _Atomic_wait_on_address_statically_avaliable = true;
+#else
+inline constexpr bool _Atomic_wait_on_address_statically_avaliable = false;
+#endif
+}
+
+inline bool __std_atomic_wait_spin(unsigned long long& _Wait_context, const bool _Is_direct) noexcept {
+    if constexpr (_Atomic_wait_on_address_statically_avaliable) {

```

Comment:
Hmmm if constexpr in a non-template is a bit "new" for us because normally that's only necessary to guard ill formed things, and being not in a template means it doesn't actually guard that. Part of me wants to acknowledge that we are doing effectively a preprocessor thing and would want to stay in the preprocessor domain, but I do like avoiding the preprocessor wherever possible... I'm torn.

Either way there probably needs to be a (void) cast to _Is_direct because c1xx will warn about unused variable when _Atomic_wait_on_address_statically_avaliable is false.

Reply:
"Compiler throughput" towards preprocessor. On the other hand, non-ill-formed both branch is useful for maintenance (don't have to build both branches).
Since STL should try to be more friendly to users rather than to itself, preprocessor looks more suitable option, I'll change that.

Reply:
> Either way there probably needs to be a (void) cast to _Is_direct because c1xx will warn about unused variable when _Atomic_wait_on_address_statically_avaliable is false.

No. It is passed to `__std_atomic_get_spin_count`, and then runtime check for direct `WaitOnAddress` is performed in .cpp file.